packages/metamask/src/index.ts

```diff
@@ -133,42 +137,40 @@ export class MetaMask extends Connector {
       .then(async () => {
         if (!this.provider) throw new NoMetaMaskError()
 
-        return Promise.all([
-          this.provider.request({ method: 'eth_chainId' }) as Promise<string>,
-          this.provider.request({ method: 'eth_requestAccounts' }) as Promise<string[]>,
-        ]).then(([chainId, accounts]) => {
-          const receivedChainId = parseChainId(chainId)
-          const desiredChainId =
-            typeof desiredChainIdOrChainParameters === 'number'
-              ? desiredChainIdOrChainParameters
-              : desiredChainIdOrChainParameters?.chainId
-
-          // if there's no desired chain, or it's equal to the received, update
-          if (!desiredChainId || receivedChainId === desiredChainId)
-            return this.actions.update({ chainId: receivedChainId, accounts })
-
-          const desiredChainIdHex = `0x${desiredChainId.toString(16)}`
-
-          // if we're here, we can try to switch networks
-          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-          return this.provider!.request({
-            method: 'wallet_switchEthereumChain',
-            params: [{ chainId: desiredChainIdHex }],
-          })
-            .catch((error: ProviderRpcError) => {
-              if (error.code === 4902 && typeof desiredChainIdOrChainParameters !== 'number') {
-                // if we're here, we can try to add a new network
-                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-                return this.provider!.request({
-                  method: 'wallet_addEthereumChain',
-                  params: [{ ...desiredChainIdOrChainParameters, chainId: desiredChainIdHex }],
-                })
-              }
-
-              throw error
-            })
-            .then(() => this.activate(desiredChainId))
+        // Wallets may resolve eth_chainId and hang on eth_accounts pending user interaction, which may include changing
+        // chains; they should be requested serially, with accounts first, so that the chainId can settle.
+        const accounts = await this.provider.request({ method: 'eth_requestAccounts' }) as string[]
```

Comment:
follow-up to my previous question: why no fallback to `eth_accounts` here?

Reply:
MetaMask is known to implement this method, whereas a pure EIP1193 may not.